src/types/dict.rs

```diff
@@ -157,19 +157,17 @@ impl PyDict {
     /// returns `Ok(None)` if item is not present, or `Err(PyErr)` if an error occurs.
     ///
     /// To get a `KeyError` for non-existing keys, use `PyAny::get_item_with_error`.
-    #[cfg(not(PyPy))]
     pub fn get_item_with_error<K>(&self, key: K) -> PyResult<Option<&PyAny>>
     where
         K: ToPyObject,
     {
         unsafe {
             let ptr =
                 ffi::PyDict_GetItemWithError(self.as_ptr(), key.to_object(self.py()).as_ptr());
-            if !ffi::PyErr_Occurred().is_null() {
-                return Err(PyErr::fetch(self.py()));
+            match NonNull::new(ptr) {
+                Some(p) => Ok(Some(self.py().from_owned_ptr(ffi::_Py_NewRef(p.as_ptr())))),
+                None => PyErr::take(self.py()).map(Err).transpose(),
```

Comment:
This code is very dense. Maybe it would help to at least take `let py = self.py();`? Do we have some kind of helper method for `from_owned_ptr(_Py_NewRef(ptr))`? Isn't this `from_borrowed

Reply:
`py.from_borrowed_ptr` doesn't actually increase the reference count, it just casts to the `&PyAny`. Pretty easy to misuse tbh, so I'll be glad to see that removed with the pool.

On the other hand, `Py::from_borrowed_ptr` and equivalents _do_ convert the borrowed ptr into an owned one, because `Py` is an owning type. So I'll switch to use that.

Reply:
Ok, so as we are going all-out code golf on this one, I think combinators and a single transpose would work as well, i.e.

```rust
 PyObject::from_borrowed_ptr_or_opt(py, ptr)
                .map(|pyobject| Ok(pyobject.into_ref(py)))
                .or_else(|| PyErr::take(py).map(Err))
                .transpose()
```

Reply:
Great spot - I had a feeling the final `.transpose()` was possible but missed what to do with the `Some` variant.